### PR TITLE
[#1013] - Enable more granular control over logging behavior

### LINKED ISSUE #1013: Allow custom logger or provide way to disable logging altogether
## Is your feature request related to a problem? Please describe.
Currently, the SDK logs internal events using its own, internal logger. The documentation describes the ability to customize this logger by passing an object conforming to the `Logger` protocol to the `Initialization` object. (see [readme](https://github.com/zcash/ZcashLightClientKit#integrating-with-logging-tools) However, the SDK currently only provides a `convenience init()` that allows users of the SDK to pass in a log level, not a custom instance of `Logger`. Aside from this being confusing, it also means that the SDK doesn't actually allow customization of the logger beyond the log level.

## Describe the solution you'd like
My use case is to disable logging altogether. To that end:

I'd like for `Initializer` struct to provide an additional `init` function such as the following:

```
convenience public init(
        cacheDbURL: URL?,
        fsBlockDbRoot: URL,
        dataDbURL: URL,
        pendingDbURL: URL,
        endpoint: LightWalletEndpoint,
        network: ZcashNetwork,
        spendParamsURL: URL,
        outputParamsURL: URL,
        saplingParamsSourceURL: SaplingParamsSourceURL,
        alias: ZcashSynchronizerAlias = .default,
        logger: Logger? = nil
) { ... }
```
For a user who doesn't care about customizing the logger, they can omit the logger parameter, in which case the SDK will default to its own, internal logger that it is using today.

For a user who wants to provide custom functionality to their logger, they can implement the protocol and pass in the conforming object. The SDK will use that to perform all logging.

For a user (such as myself) who wants to disable logging entirely, I can implement the Logger protocol and leave all the methods empty.

## Alternatives you've considered
Alternatively, since I simply would like the ability to disable logging, one approach to that could be to have an initializer that accepts a flag to determine whether logging is enabled or disabled. If this alternative is chosen as the preferred approach, I think we should update the README to remove any mention of passing in a custom logger.

### PR DESCRIPTION
**Pull request by tw0po1nt**
The `Initializer` struct now takes a `LoggingPolicy` parameter, which enables the following options when initializing the SDK:

- `default` - the SDK will default to using its own internal implementation of the `Logger` protocol at the `debug` log level. You can also override the log level by passing your preferred log level in to the enum case.
- `custom` - the user provides an object conforming to the `Logger` protocol, allowing for complete customizability of the logging.
- `noLogging` - the SDK will not log anything

Closes #1013 

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [x] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Initializer.swift [hunk 1]
```diff
@@ -166,7 +172,16 @@ public class Initializer {
         // from constructor. So `parsingError` is just stored in initializer and `SDKSynchronizer.prepare()` throw this error if it exists.
         let (updatedURLs, parsingError) = Self.tryToUpdateURLs(with: alias, urls: urls)
 
-        let logger = OSLogger(logLevel: logLevel, alias: alias)
+        var logger: Logger
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Initializer.swift:175]: This variable can be `let`. It is just tiny detail.

File: Sources/ZcashLightClientKit/Utils/LoggingProxy.swift [hunk 2]
```diff
@@ -43,3 +43,14 @@ extension Logger {
         error(message, file: file, function: function, line: line)
     }
 }
+
+/**
+A concrete logger implementation that logs nothing at all
+ */
+struct NullLogger: Logger {
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Utils/LoggingProxy.swift:50]: I think it makes sense to make this struct public with it's all methods.

**Review by Chlup — CHANGES_REQUESTED** [review]

**Comment by Chlup** [comment]
I have two small comments otherwise this looks ok.

**Review by LukasKorba — APPROVED** [review]

**Comment by tw0po1nt** [comment]
@Chlup @LukasKorba 
Since this is my first PR in this repo, I just have a couple questions. 

- Do you guys usually have the PR opener merge, or do you have another maintainer manage that?
- In the contributing docs, it mentions to keep the change log up to date, but it doesn't look like some of the other open PRs have updated that. Would you like me to update that prior to merging this?

**Comment by Chlup** [comment]
> @Chlup @LukasKorba Since this is my first PR in this repo, I just have a couple questions.
> 
> * Do you guys usually have the PR opener merge, or do you have another maintainer manage that?
> * In the contributing docs, it mentions to keep the change log up to date, but it doesn't look like some of the other open PRs have updated that. Would you like me to update that prior to merging this?

Hi Matthew,

PR opener usually merges the PR after it's approved by some maintainer. Changelog is usually updated for the PRs that change public API. Which this does. Please update `CHANGELOG.md` file accordingly.

Before merging the PR please make sure that all the targets can be compiled. You can use `All` scheme for that. And please be sure that tests work. Unfortunately CI currently doesn't run all the tests. Running the `DarksideTests` can be little challenging. But we can help with that.

Have a nice day.
Michal

**Comment by Chlup** [comment]
I checked the tests and all work fine. Just please fix the compilation of the sample app and you should be good to go.

**Review by pacu — APPROVED** [review]

**Review by Chlup — APPROVED** [review]

**Pull request merged**